nullaway/src/main/java/com/uber/nullaway/handlers/AbstractFieldContractHandler.java

```diff
@@ -0,0 +1,247 @@
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.handlers;
+
+import com.google.common.base.Preconditions;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Symbol;
+import com.uber.nullaway.NullAway;
+import com.uber.nullaway.NullabilityUtil;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+
+/**
+ * Abstract base class for any annotation which aims processing class fields to satisfy pre/post
+ * conditions. Note: all fields that are going to be processed must be the fields of the receiver.
+ * (e.g. field or this.field)
+ */
+public abstract class AbstractFieldContractHandler extends BaseNoOpHandler {
+
+  protected static final String THIS_NOTATION = "this.";
+  /** Simple name of the annotation in {@code String} */
+  protected String annotName;
+
+  /**
+   * This method verifies that the method adheres to any corresponding
+   * (@EnsuresNonNull/@RequiresNonNull) annotation.
+   */
+  @Override
+  public void onMatchMethod(
+      NullAway analysis, MethodTree tree, VisitorState state, Symbol.MethodSymbol methodSymbol) {
+    Set<String> annotationContent = getFieldNamesFromAnnotation(methodSymbol);
+    boolean isValid =
+        validateAnnotationSyntax(annotationContent, analysis, tree, state, methodSymbol);
+    isValid =
```

Comment:
This logic is still pretty convoluted.  I think you can write as a single boolean condition:
```java
boolean isValid = annotationContent != null 
       && validateAnnotationSyntax(annotationContent, analysis, tree, state, methodSymbol)
       && validateAnnotationSemantics(analysis, state, tree, methodSymbol);
```
Is there any problem with that?

Reply:
Actually this will lead to an error when checking overriding rules. I added a new ```boolean``` variable with name ```isAnnotated``` to make it more readable.

```java
Set<String> annotationContent = getFieldNamesFromAnnotation(methodSymbol);
    boolean isAnnotated = annotationContent != null;
    boolean isValid =
        isAnnotated
            && validateAnnotationSyntax(annotationContent, analysis, tree, state, methodSymbol)
            && validateAnnotationSemantics(analysis, state, tree, methodSymbol);
    if (isAnnotated && !isValid) {
      return;
    }
    Symbol.MethodSymbol closestOverriddenMethod =
        NullabilityUtil.getClosestOverriddenMethod(methodSymbol, state.getTypes());
    if (closestOverriddenMethod == null) {
      return;
    }
    Set<String> fieldNames;
    if (isAnnotated) {
      fieldNames =
          annotationContent
              .stream()
              .map(AbstractFieldContractHandler::trimReceiver)
              .collect(Collectors.toSet());
    } else {
      fieldNames = Collections.emptySet();
    }
    validateOverridingRules(fieldNames, analysis, state, tree, closestOverriddenMethod);
    super.onMatchMethod(analysis, tree, state, methodSymbol);
```